ledger-api/grpc-definitions/com/daml/ledger/api/v1/event_query_service.proto

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) 2023 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+syntax = "proto3";
+
+package com.daml.ledger.api.v1;
+
+import "com/daml/ledger/api/v1/event.proto";
+import "com/daml/ledger/api/v1/value.proto";
+
+option java_outer_classname = "EventQueryServiceOuterClass";
+option java_package = "com.daml.ledger.api.v1";
+option csharp_namespace = "Com.Daml.Ledger.Api.V1";
+
+// Allow clients to query events
+service EventQueryService {
+
+  // Get the create and the consuming exercise event for the contract with the provided ID
+  rpc GetEventsByContractId (GetEventsByContractIdRequest) returns (GetEventsByContractIdResponse);
+
+  // Get all create and consuming exercise events for the contracts with the provided contract key
+  rpc GetEventsByContractKey (GetEventsByContractKeyRequest) returns (GetEventsByContractKeyResponse);
+
+}
+
+message GetEventsByContractIdRequest {
+  // The contract id being queried.
+  // Required
+  string contract_id = 1;
+
+  // The parties whose events the client expects to see.
+  // The events associated with the contract id will only be returned if the requesting parties includes
+  // at least one party that is a witness of the action that created the event. For a full description
+  // of witness see https://docs.daml.com/concepts/ledger-model/ledger-privacy.html#projections.
+  // Required
+  repeated string requesting_parties = 2;
+
+}
+
+message GetEventsByContractIdResponse {
+
+  // Will be populated if a created event is found for the contract id
+  // Optional
+  CreatedEvent create_event = 1;
+
+  // Will be populated if a archive event is found for the contract id
```

Comment:
Given that you're filtering by witnesses rather than doing stakeholder filtering I can get both only the create and only the archive event right? Probably will confuse some people but seems like the right thing to do. The one thing we could consider is if we should include the contract payload in the ArchivedEvent since technically I am being divulged that but that's a somewhat orthogonal change so probably reasonable to not include it for now.

Reply:
I was copying the behaviour of the transaction stream. On consideration I think switching to stakeholders is a good idea for the reasons you outline here and below.